### PR TITLE
Stop desktop app window's QML source from running when window is closed

### PR DESCRIPTION
**Pull request by ctrlaltdavid**

### CODE REVIEW

**Comment by ctrlaltdavid** [comment]
### Test Plan


#### Preparations

1) Edit the PlotPerf.qml file so that it repeatedly prints output to the program log:
- Find PlotPerf.qml in the PR's install directory under /scripts/developer/utilities/lib/plotperf/
- Open the file in a text editor running with administrator privileges (so that you can save your edits).
- Find the "onPaint" code block that starts with the following line:
```
    onPaint: {
```
- Immediately after this line, add a line that prints output to the program log. For example: such as the following after this line
```
    onPaint: {
        print("@@@@@@@ onPaint");
```

2) Check that output is repeatedly printed to the program log.
- Run Interface, and search for and load the script "lod.js" (is in /developer/utilities/render/). An app icon labelled "LOD" is added to the toolbar / tablet.
- Open the program log viewer, Developer > Log.
- Desktop mode.
- Click on the "LOG" icon. A dialog should display graphs of graphics information and  "@@@@@@@ onPaint" should be being repeatedly output to the program log.


#### Desktop

Per 1) and 2), check that you're in desktop mode with the LOD dialog open, then ...

3) Click on the "LOD" icon. The dialog should close and the print output should stop being output to the program log.

4) Open the LOD dialog, then click on another toolbar button that opens another dialog. The LOD dialog should close, the print output should stop being output to the program log, and the new dialog should open.


#### HMD

Change to HMD mode, then ...

5) Open the LOD app. The print output should start being output to the program log. Click the tablet "x" to close the app and return to the main menu. The print output should stop being output to the program log.

6) Open the LOD app. The print output should start being output to the program log. Click the controller button to hide the tablet. The print output should stop being output to the program log. Click the controller button again to show the tablet. The LOD app should resume from where it left off and the print output should start being output to the program log again.

https://highfidelity.manuscript.com/f/cases/16039

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10756/HighFidelity-Beta-PR13427-764e00d.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10756/HighFidelity-Beta-PR13427-764e00d.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10756/HighFidelity-Beta-Interface-PR13427-764e00d.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10756/HighFidelity-Beta-PR13427-764e00d.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10756/HighFidelity-Beta-Interface-PR13427-764e00d.exe

File: libraries/qml/src/qml/OffscreenSurface.cpp [hunk 1]
```diff
@@ -286,6 +286,13 @@ void OffscreenSurface::loadInternal(const QUrl& qmlSource,
     if (QThread::currentThread() != thread()) {
         qFatal("Called load on a non-surface thread");
     }
+
+    // For desktop toolbar mode window: stop script when window is closed.
+    if (qmlSource.isEmpty()) {
+        getSurfaceContext()->engine()->quit();
+        return;
+    }
```
Comment (danteruiz) [inline libraries/qml/src/qml/OffscreenSurface.cpp]: Not sure what this does, doesn't look right. Can you explain to me what this does.

Comment (ctrlaltdavid) [reply libraries/qml/src/qml/OffscreenSurface.cpp]: In desktop toolbar mode, the QML in a QML tablet app isloaded into a persistent _desktopWindow object and its QML starts running. The _desktopWindow object is not deleted when the dialog is closed (via `"setShown"` message with `false` parameter). The `"setShown"` just hides the window but its QML keeps running.
Hence the `stopQMLSource()` call added after invoking `"setShown"` which clears the current QML source to `""`, and upon this being loaded in `OffscreenSurface::loadInternal()` the current QML script engine is stopped. (Without this call the QML's script engine keeps running and in the QA test would the "@@@@@@@ onPaint" would keep printing to the program log.)

Comment (danteruiz) [reply libraries/qml/src/qml/OffscreenSurface.cpp]: so this destroys the qml engine?

Comment (ctrlaltdavid) [reply libraries/qml/src/qml/OffscreenSurface.cpp]: No, it stops the engine from running (as evidenced by the debug messages not longer being output). The engine is still there, ready for the QML for another desktop dialog to be loaded.

**Comment by qliacoombs** [comment]
Build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/10928/HighFidelity-Beta-PR13427-764e00d.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10928/HighFidelity-Beta-PR13427-764e00d.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/10928/HighFidelity-Beta-Interface-PR13427-764e00d.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10928/HighFidelity-Beta-PR13427-764e00d.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/10928/HighFidelity-Beta-Interface-PR13427-764e00d.exe

**Comment by qlimnavarro** [comment]
Pass @conklin94122 
Script log appears as expected.

**Pull request merged**